### PR TITLE
Internal: [NA] Tests task - Create E2E tests [ED-11088]

### PR DESCRIPTION
**Pull request by Barcov**

### CODE REVIEW

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js [hunk 1]
```diff
@@ -149,48 +149,64 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
...
 				//  Verify that all items are closed.
 				await editor.openSection( 'section_interactions' );
 				await editor.setSelectControlValue( 'default_state', 'all_collapsed' );
 
 				for ( const item of allItems ) {
-					await expect( item ).not.toHaveAttribute( 'open', '' );
+					await expect.soft( item ).not.toHaveAttribute( 'open', '' );
 				}
 
 				// Check manual select of first expand -> first item is open
```
Comment (TzviRabinovitch) [inline tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:189]: maybe use test.step instead of comments?

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js [hunk 2]
```diff
@@ -149,48 +149,64 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
...
-					secondItemID = await page.locator( '.e-n-accordion-item >> nth=1' ).first().getAttribute( 'id' ),
-					thirdItemID = await page.locator( '.e-n-accordion-item >> nth=2' ).first().getAttribute( 'id' );
+				for ( const item of allItemsExceptFirst ) {
+					await expect.soft( item ).not.toHaveAttribute( 'open', 'true' );
+				}
 
-				await expect( firstItemID ).not.toEqual( secondItemID );
-				await expect( secondItemID ).not.toEqual( thirdItemID );
-				await expect( thirdItemID ).not.toEqual( firstItemID );
+				// Reset max_items_expanded to default state
```
Comment (TzviRabinovitch) [inline tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:208]: not sure if the comment is needed here

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js [hunk 3]
```diff
@@ -253,16 +269,76 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
...
+			} );
+		} );
+
+		test( 'Nested Accordion test SVG Icon and No Icon', async ( { browser }, testInfo ) => {
+			const page = await browser.newPage(),
+				wpAdmin = new WpAdminPage( page, testInfo ),
+				nestedAccordionWidgetId = '48f02ad';
+
+			await wpAdmin.enableAdvancedUploads();
+			const editor = await wpAdmin.useElementorCleanPost();
```
Comment (TzviRabinovitch) [inline tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:287]: useElementorCleanPost might prevent parallel execution ( not used yet )

File: tests/playwright/pages/editor-page.js [hunk 4]
```diff
@@ -778,4 +794,34 @@ module.exports = class EditorPage extends BasePage {
 		await this.setSliderControlValue( selector + '_font_size', fontsize );
 		await this.page.locator( '.elementor-control-' + selector + '_typography .eicon-edit' ).click();
 	}
+
+	/*
+	* Checks for a stable UI state by comparing screenshots at intervals and expecting a match.
+	* Can be used to check for completed rendering. Useful to wait out animations before screenshots and expects.
+	* Should be less flaky than waitForLoadState( 'load' ) in editor where Ajax re-rendering is triggered.
+	*/
+	async isScreenShottable( locator, retries = 3, timeout = 500 ) {
```
Comment (louiswol94) [inline tests/playwright/pages/editor-page.js:803]: @Barcov Nice!

**Review by TzviRabinovitch — DISMISSED** [review]

**Comment by hein-obox** [comment]
@Barcov 
Would it be possible to update `loadJsonPageTemplate` and remove the code duplication from `loadTemplate`  and to call the `loadTemplate` inside the `loadJsonPageTemplate` function?

**Review by davseve — DISMISSED** [review]

File: tests/playwright/pages/editor-page.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Barcov) [reply tests/playwright/pages/editor-page.js:803]: @davseve I like the proposal for broader naming. I've noticed that I started using this to check stability before other expects than Screenshots. Extracting this would be useful for FE as well.

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js [hunk 5]
```diff
@@ -355,146 +436,145 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
...
+				await setTitleIconPosition( 'left', editor, 'mobile' );
 				// Assert
-				await frame.waitForLoadState( 'load' );
-				await expect( nestedAccordionTitle.locator( '.e-n-accordion-item-title-icon' ) ).toHaveCSS( 'order', '-1' );
+				await editor.isScreenShottable( nestedAccordionTitle, 3 );
+				await expect.soft( nestedAccordionTitle.locator( '.e-n-accordion-item-title-icon' ) ).toHaveCSS( 'order', '-1' );
 			} );
 		} );
 
 		test( 'Nested Accordion Title Icon and Text Vertical Alignment', async ( { browser }, testInfo ) => {
```
[the remark this answers is not in the document]
Comment (Barcov) [reply tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:500]: @davseve 3 in this case refers to the number of retries.

```
async isScreenShottable( locator, retries = 3, timeout = 500 ) {
```

File: tests/playwright/pages/editor-page.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (hein-obox) [reply tests/playwright/pages/editor-page.js:803]: @Barcov Please send me a PM about this, because I have copied and pasted your function into the pro version as well :).

[the remark this answers is not in the document]
Comment (hein-obox) [reply tests/playwright/pages/editor-page.js:803]: Btw It would be so much better if we would have one testing codebase.

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js [hunk 6]
```diff
@@ -253,16 +272,77 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
...
+				nestedAccordionWidgetId = '48f02ad';
+
+			await wpAdmin.enableAdvancedUploads();
+			const editor = await wpAdmin.useElementorCleanPost();
+			let	frame = editor.getPreviewFrame();
+
+			await editor.loadJsonPageTemplate( __dirname, 'nested-accordion-title-and-icons', '.elementor-widget-n-accordion' );
+			await editor.closeNavigatorIfOpen();
+
+			await test.step( 'Check that an SVG title icon is displayed', async () => {
```
[the remark this answers is not in the document]
Comment (Barcov) [reply tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:296]: @davseve We can do that. Drop me a PM and let's discuss the details.

**Comment by Barcov** [comment]
> @Barcov Would it be possible to update `loadJsonPageTemplate` and remove the code duplication from `loadTemplate` and to call the `loadTemplate` inside the `loadJsonPageTemplate` function?

@hein-obox I think we'll have to push this refactoring out to another task to compare pro and core branches and reconcile differences in a next sprint.

File: tests/playwright/pages/editor-page.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Barcov) [reply tests/playwright/pages/editor-page.js:803]: @hein-obox & @davseve We'll need to consider this refactor in a new task next sprint.

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Barcov) [reply tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:296]: @davseve On testing using SVG icons depends on inline_font experiment being enabled.

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js (hunk 3 above)
Comment (Barcov) [reply tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js]: Good idea @TzviRabinovitch, I've removed all usage of this function in NA tests.

File: tests/playwright/sanity/modules/nested-accordion/helper.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+ * Take a Screenshot of this Widget
+ *
+ * @param {string} fileName
+ * @param {Object} locator
+ * @return {Promise<void>}
+ */
+async function expectScreenshotToMatchLocator( fileName, locator ) {
+	expect.soft( await locator.screenshot( {
+		type: 'png',
+	} ) ).toMatchSnapshot( fileName );
```
Comment (louiswol94) [inline tests/playwright/sanity/modules/nested-accordion/helper.js:29]: @Barcov Is this test flaky? It failed once.

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 8]
```diff
@@ -23,6 +23,7 @@ yarn.lock
 .project/
 .vscode/
 .idea/
+/.run/
```
[the remark this answers is not in the document]
Comment (Barcov) [reply .gitignore:26]: PhpStorm settings for local runners like grunt
image

File: .run/watch-scripts.run.xml [hunk 9]
```diff
@@ -1,11 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
[the remark this answers is not in the document]
Comment (Barcov) [reply .run/watch-scripts.run.xml:1]: @louiswol94 This is a Dev config for local runners like grunt in PhpStorm

File: .run/watch-styles.run.xml [hunk 10]
```diff
@@ -1,11 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
[the remark this answers is not in the document]
Comment (Barcov) [reply .run/watch-styles.run.xml:1]: @louiswol94 As above

File: tests/playwright/pages/editor-page.js [hunk 11]
```diff
@@ -544,6 +560,34 @@ module.exports = class EditorPage extends BasePage {
...
+	async editCurrentPage() {
+		const postId = await this.getPageID();
+		expect( postId, 'No Post/Page ID returned when calling getPageId(). This may indicate unexpected permalink settings - try using wpAdmin.setPermalink()' ).toBeTruthy();
+
+		if ( null !== postId ) {
+			await this.gotoPostId( postId );
+		}
+	}
+
+	async getPageID() {
```
[the remark this answers is not in the document]
Comment (Barcov) [reply tests/playwright/pages/editor-page.js:572]: @louiswol94 Yes David and I where discussing this and I have a change like that to merge in on a seperate PR.

File: .gitignore [hunk 12]
```diff
@@ -34,3 +35,4 @@ coverage/
 test-results/
 **/storageState.json
 tests/playwright/config/local.config.js
+
```
[the remark this answers is not in the document]
Comment (Barcov) [reply .gitignore:38]: @louiswol94 Storm change. Reverting...

File: tests/playwright/sanity/modules/nested-accordion/helper.js (hunk 7 above)
Comment (Barcov) [reply tests/playwright/sanity/modules/nested-accordion/helper.js:29]: @louiswol94 This is essentially shorthand for a normal screenshot with png and soft expect dialled in. So could only be flaky if the locator is flaky. 

We'll possibly need to increase the retires for isUIStable() on the failing screenshot if we see this again.

File: modules/nested-accordion/widgets/nested-accordion.php [hunk 13]
```diff
@@ -767,20 +766,34 @@ protected function content_template() {
...
+			});
+
+			view.addRenderAttribute( itemTitleText, {
+				'class': ['e-n-accordion-item-title-text'],
+				'data-binding-type': 'repeater-item',
+				'data-binding-repeater-name': 'items',
+				'data-binding-setting': ['item_title'],
+				'data-binding-index': itemCount,
+			});
+
```
[the remark this answers is not in the document]
Comment (hein-obox) [reply modules/nested-accordion/widgets/nested-accordion.php]: @Barcov Shouldn't this be inside another PR?

File: modules/nested-accordion/widgets/nested-accordion.php [hunk 14]
```diff
@@ -271,7 +271,6 @@ protected function register_controls() {
 				],
 				'default' => 'div',
 				'separator' => 'before',
-				'render_type' => 'template',
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply modules/nested-accordion/widgets/nested-accordion.php]: @hein-obox No, the fix for that would be the data-bindings on the title text. Barco removed this key on the title tag which is incorrect.

File: .run/watch-scripts.run.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (louiswol94) [reply .run/watch-scripts.run.xml:1]: @Barcov Please revert this change. Should not be in this PR :-)

File: .run/watch-styles.run.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (louiswol94) [reply .run/watch-styles.run.xml:1]: @Barcov Please revert this change. Should not be in this PR :-)

File: modules/nested-accordion/widgets/nested-accordion.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (louiswol94) [reply modules/nested-accordion/widgets/nested-accordion.php]: @Barcov Please revert this change. Should not be in this PR :-)

File: modules/nested-accordion/widgets/nested-accordion.php [hunk 15]
```diff
@@ -808,7 +821,7 @@ protected function content_template() {
 			<details {{{ view.getRenderAttributeString( itemWrapperKey ) }}}>
 				<summary {{{ view.getRenderAttributeString( itemTitleKey ) }}}>
 					<span class="e-n-accordion-item-title-header">
-						<{{{ titleHTMLTag }}} class="e-n-accordion-item-title-text">
+						<{{{ titleHTMLTag }}} {{{ view.getRenderAttributeString( itemTitleText ) }}}>
```
Comment (louiswol94) [inline modules/nested-accordion/widgets/nested-accordion.php]: @Barcov Please revert this change. Should not be in this PR :-)

File: tests/playwright/pages/editor-page.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (louiswol94) [reply tests/playwright/pages/editor-page.js:572]: @Barcov I think you should just update it in this function already.

File: tests/playwright/pages/wp-admin-page.js [hunk 16]
```diff
@@ -166,4 +171,22 @@ module.exports = class WpAdminPage extends BasePage {
 	async editWithElementor() {
 		await this.page.getByRole( 'link', { name: ' Edit with Elementor' } ).click();
 	}
+
+	async setPermalinks( structure = 'plain' ) {
```
Comment (hein-obox) [inline tests/playwright/pages/wp-admin-page.js:175]: Nice!

File: tests/playwright/sanity/modules/nested-accordion/helper.js (hunk 7 above)
Comment (hein-obox) [inline tests/playwright/sanity/modules/nested-accordion/helper.js:29]: What is causing the problem?

**Review by hein-obox — DISMISSED** [review]

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion-style.test.js [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+		} );
+
+		await test.step( 'Container\'s style should override item\'s style', async () => {
+			await test.step( 'Open container settings', async () => {
+				// Act
+				await nestedAccordionItemContent.first().hover();
+				await nestedAccordionItemTitle.first().locator( '.elementor-editor-container-settings' ).click();
+			} );
+
+			await test.step( 'override background and border', async () => {
```
Comment (hein-obox) [inline tests/playwright/sanity/modules/nested-accordion/nested-accordion-style.test.js:130]: I would expect this to start with a capital letter (Here and everywhere).

File: tests/playwright/sanity/modules/nested-accordion/nested-accordion-style.test.js [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+				} );
+				await page.waitForTimeout( 1000 );
+
+				// Assert.
+				await editor.isUiStable( nestedAccordionWidgetFront );
+				await expectScreenshotToMatchLocator( 'nested-Accordion-content-style-front.png', nestedAccordionWidgetFront );
+			} );
+		} );
+	} );
+	test( 'Header style tests', async ( { page }, testInfo ) => {
```
Comment (hein-obox) [inline tests/playwright/sanity/modules/nested-accordion/nested-accordion-style.test.js:173]: I would expect an empty line above the `test()`. Here and everywhere.

File: tests/playwright/sanity/modules/nested-accordion/helper.js (hunk 7 above)
Comment (Barcov) [reply tests/playwright/sanity/modules/nested-accordion/helper.js:29]: @hein-obox  To fix the issue with "exception require container containers" I had to abandon using selectElement() for this particular test. Dymitro suggested we replace this because it's flaking out and causing those console exceptions "container containers required".

Clicking on the widget by ID resulted in the inner container getting selected which makes the test fail.
Workaround was to click on the item-title. But this closes the first Item. So the test failed once after pushing up this change. 

I'm not expecting this to be flaky. The failure here is because of different height like the diff screenshot I added below.

image

File: tests/playwright/pages/editor-page.js [hunk 19]
```diff
@@ -544,6 +559,24 @@ module.exports = class EditorPage extends BasePage {
...
+		await this.gotoPostId( postId );
+	}
+
+	async getPageId() {
+		const pageId = await this.page.evaluate( () => elementor.config.initial_document.id );
+		return pageId;
+	}
+
+	async getPageIdFromFrontEnd() {
+		const pageId = await this.page.evaluate( () => elementorFrontendConfig.post.id );
```
Comment (louiswol94) [inline tests/playwright/pages/editor-page.js:577]: @Barcov Should `pageId` not be returned?

Comment (davseve) [inline tests/playwright/pages/editor-page.js:577]: I think we can unite these two functions into one,
 `elementorFrontendConfig.post.id` is also available in the editor. 
```suggestion
	async getPageId() {
		const pageId = await this.page.evaluate( () => elementorFrontendConfig.post.id );
		return pageId;
	}
		
```

Comment (Barcov) [reply tests/playwright/pages/editor-page.js]: @louiswol94 Nice save, I've combined getPageIdFromFrontEnd() with getPageId() and included the return

**Review by hein-obox — DISMISSED** [review]

File: tests/playwright/pages/editor-page.js [hunk 20]
```diff
@@ -544,6 +559,24 @@ module.exports = class EditorPage extends BasePage {
...
+	async editCurrentPage() {
+		const postId = await this.getPageIdFromFrontEnd();
+		await expect( postId, 'No Post/Page ID returned when calling getPageIdFromFrontEnd().' ).toBeTruthy();
+		await this.gotoPostId( postId );
+	}
+
+	async getPageId() {
+		const pageId = await this.page.evaluate( () => elementor.config.initial_document.id );
+		return pageId;
+	}
```
Comment (louiswol94) [inline tests/playwright/pages/editor-page.js:574]:
```suggestion
	async getPageId() {
		const pageId = await this.page.evaluate( () => elementor.config.initial_document.id );
		return pageId;
	}

```

File: tests/playwright/pages/wp-admin-page.js (hunk 16 above)
Comment (louiswol94) [reply tests/playwright/pages/wp-admin-page.js:175]: @Barcov This function is not used anywhere. Do we need it?

File: tests/playwright/sanity/modules/nested-accordion/helper.js [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+		right: '.eicon-h-align-right',
+		left: '.eicon-h-align-left',
+	} );
+
+	const controlBreakpoint = ( breakpoint.toLowerCase() !== 'desktop' ) ? '_' + breakpoint : '';
+	const locator = await getChoicesButtonSelector( 'item_title_icon_position' + controlBreakpoint, icon[ direction ] );
+	await editor.page.locator( locator ).click();
+}
+async function setTitleHorizontalAlignment( direction, editor, breakpoint = 'desktop' ) {
+	const icon = Object.freeze( {
```
Comment (louiswol94) [inline tests/playwright/sanity/modules/nested-accordion/helper.js:47]:
```suggestion
	await editor.page.locator( locator ).click();
}

async function setTitleHorizontalAlignment( direction, editor, breakpoint = 'desktop' ) {
	const icon = Object.freeze( {
```

File: tests/playwright/sanity/modules/nested-accordion/helper.js [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+
+	async function setBorderColor() {
+		await editor.setColorControlValue( borderColor, 'accordion_border_' + state + '_color' );
+	}
+
+	async function setState() {
+		await editor.page.click( '.elementor-control-accordion_' + state + '_border_and_background' );
+	}
+}
+async function setIconColor( editor, state, color, context ) {
```
Comment (louiswol94) [inline tests/playwright/sanity/modules/nested-accordion/helper.js:89]:
```suggestion
}

async function setIconColor( editor, state, color, context ) {
```

File: tests/playwright/pages/wp-admin-page.js (hunk 16 above)
Comment (Barcov) [reply tests/playwright/pages/wp-admin-page.js:175]: @louiswol94 I used it initially in this PR. We'll be using this in Taxonomy filter soon. It's also useful to tests that Hein has written in Mega menu. So I've left it in after refactoring I can remove it and add it in a separate PR at some point if you prefer...

Comment (Barcov) [reply tests/playwright/pages/wp-admin-page.js]: @louiswol94 removing form this PR. Refactoring has orphaned it. I'll add it in again on a later PR.

**Review by louiswol94 — APPROVED** [review]
